docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+
+For performance and flexibility for such a project, we adhere to few techniques that may seem unconventional:
+* Data-oriented Design
+* Entity Component System
+
+## Data-oriented Design (DoD)
+
+Data-oriented Design is an emerging approach to software architecture, used extensively in OSP-Magnum. By focusing on creating a memory layout for efficient access, DoD greatly reduces complexity and achieves high performance. As counterintuitive as it may seem, this potentially results in (or at least encourages) a greater degree of flexibility compared to class-based OOP.
+
+DoD is not a new idea, but it is incredibly unconventional to most programmers (especially ones adept to OOP). It will be useful to read about or listen to a few presentations before getting started.
+
+See [data-oriented-design](https://github.com/dbartolini/data-oriented-design) for plenty of resources.
+
+Important note: When going down the rabbit hole, *Data-oriented programming* and *Data-driven design* is not what you're looking for. *Correct* resources on DoD usually mention the CPU cache on systems-level languages.
+
+*tl;dr: RAM sucks at random access and use arrays everywhere because they're fast!*
+
+### DoD in practice
+
+DoD leads to:
+
+* **Seprated Data and Functions**. This idea is also favored by functional programming. Most classes are only responsible for storing data and their methods only modify their own members. Most operations are free functions that only have a few responsibilities and little side effects.
+* **Structure of Arrays**. Aside from efficiency, data can be organized based on how it's operated on. Data for a single 'object' can be represented across multiple arrays, and each of these arrays can be passed to functions that can run simultaneously. This also leads to a natural separation of concerns.
+* **Extensive use of contiguous containers**. Dealing with streams of data is the common case. Contiguous containers offer best performance with O(1) access and perfect cache locality. Can't really say `std::vector` is the best, but it is heavily used.
+* **No runtime Polymorphism**. We don't need polymorphism. Instead, prefer a more true form of message passing through shared variables. Two completely different systems can communicate using queues that act as interfaces for example.
+* **Less Abstractions**. This is not a bad thing, nor does it mean everything is written in assembly. Along with separate data and functions, the computer has separate RAM and CPU (Objects aren't real!). By following a style that more resembles the computer's architecture, less (unnecessary) abstractions are needed to glue together the 'perfect' interface.
+
+Note that this doesn't mean everything has to be ultra-optimized. It's perfectly acceptable to prefer a more convenient option for paths that aren't performance critical, such as operations called only once per frame or once during setup.
+
+## Entity Component System (ECS)
```

Comment:
If you're not in a rush to merge this, I may take an editing pass over this section and commit my changes; otherwise I can wait and make another PR. Since I've explained these comments so many times to people on reddit, I have some useful experience for communicating these details.